static/js/recent_topics_ui.js

```diff
@@ -252,38 +298,83 @@ export function process_messages(messages) {
-    const stream_muted = stream_data.is_muted(stream_id);
-    const muted = topic_muted || stream_muted;
-    const unread_count = unread.num_unread_for_topic(stream_id, topic);
-
-    // Display in most recent sender first order
-    const all_senders = recent_senders.get_topic_recent_senders(stream_id, topic);
-    const senders = all_senders.slice(-MAX_AVATAR);
-    const senders_info = people.sender_info_for_recent_topics_row(senders);
-
-    // Collect extra senders fullname for tooltip.
-    const extra_sender_ids = all_senders.slice(0, -MAX_AVATAR);
-    const displayed_other_senders = extra_sender_ids.slice(-MAX_EXTRA_SENDERS);
-    const displayed_other_names = people.get_display_full_names(displayed_other_senders.reverse());
+    const type = last_msg.type;
+    msg_info.full_last_msg_date_time = timerender.get_full_datetime(time);
+    msg_info.conversation_key = get_key_from_message(last_msg);
+    msg_info.unread_count = unread.get_thread_unread_count_from_message(last_msg);
+    msg_info.last_msg_time = timerender.last_seen_status_from_date(time);
+    let displayed_other_names;
+    let extra_sender_ids;
+
+    if (type === "stream") {
+        const stream_info = sub_store.get(last_msg.stream_id);
+
+        // Stream info
+        msg_info.stream_id = last_msg.stream_id;
+        msg_info.stream = last_msg.stream;
+        msg_info.stream_color = stream_info.color;
+        msg_info.stream_url = hash_util.by_stream_url(msg_info.stream_id);
+        msg_info.invite_only = stream_info.invite_only;
+        msg_info.is_web_public = stream_info.is_web_public;
+
+        // Topic info
+        msg_info.topic = last_msg.topic;
+        msg_info.topic_url = hash_util.by_stream_topic_url(msg_info.stream_id, msg_info.topic);
+
+        // We hide the row according to filters or if it's muted.
+        // We only supply the data to the topic rows and let jquery
+        // display / hide them according to filters instead of
+        // doing complete re-render.
+        msg_info.topic_muted = Boolean(
+            user_topics.is_topic_muted(msg_info.stream_id, msg_info.topic),
+        );
+        const stream_muted = stream_data.is_muted(msg_info.stream_id);
+        msg_info.muted = msg_info.topic_muted || stream_muted;
+
+        // Display in most recent sender first order
+        const all_senders = recent_senders.get_topic_recent_senders(
+            msg_info.stream_id,
+            msg_info.topic,
+        );
+        const senders = all_senders.slice(-MAX_AVATAR);
+        msg_info.senders = people.sender_info_for_recent_topics_row(senders);
+        msg_info.other_senders_count = Math.max(0, all_senders.length - MAX_AVATAR);
+
+        // Collect extra sender fullname for tooltip
+        extra_sender_ids = all_senders.slice(0, -MAX_AVATAR);
+        const displayed_other_senders = extra_sender_ids.slice(-MAX_EXTRA_SENDERS);
+        displayed_other_names = people.get_display_full_names(displayed_other_senders.reverse());
```

Comment:
I think you can probably deduplicate most of the ~6 lines here for populating the `senders` / etc. fields; only the code to compute `all_senders` and `senders` appears to be different?